File: etapa3-construcao(geracao-de-cod)/stack2/src/backend/src/main/java/com/hotel/reserva/domain/valueobjects/Email.java
package com.hotel.reserva.domain.valueobjects;

import jakarta.persistence.Embeddable;
import lombok.AccessLevel;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.NoArgsConstructor;

@Embeddable
@Getter
@EqualsAndHashCode
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class Email {

    private String valor;

    private Email(String valor) {
        this.valor = valor.toLowerCase().trim();
    }

    public static Email criar(String email) {
        if (!validar(email)) {
            throw new IllegalArgumentException("E-mail inválido");
        }
        
        return new Email(email);
    }

    private static boolean validar(String email) {
        String regex = "^[^\\s@]+@[^\\s@]+\\.[^\\s@]+$";
        return email != null && email.matches(regex);
    }
}

File: etapa3-construcao(geracao-de-cod)/stack2/src/domain/value-objects/Email.ts
export class Email {
  private readonly valor: string;

  private constructor(email: string) {
    this.valor = email.toLowerCase().trim();
  }

  static criar(email: string): Email {
    if (!this.validar(email)) {
      throw new Error('E-mail inválido');
    }
    
    return new Email(email);
  }

  private static validar(email: string): boolean {
    const regex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
    return regex.test(email);
  }

  getValor(): string {
    return this.valor;
  }

  equals(other: Email): boolean {
    return this.valor === other.valor;
  }

  toString(): string {
    return this.valor;
  }
}
